share/p2p/middleware.go

```diff
@@ -9,21 +9,35 @@ import (
 
 var log = logging.Logger("shrex/middleware")
 
-func RateLimitMiddleware(inner network.StreamHandler, concurrencyLimit int) network.StreamHandler {
-	var parallelRequests int64
-	limit := int64(concurrencyLimit)
+type Middleware struct {
+	concurrencyLimit int64
+	parallelRequests int64
+
+	// NumRateLimited is the number of requests that were rate limited. It is reset to 0 every time
+	// it is read and observed into metrics.
+	NumRateLimited int64
+}
+
+func NewMiddleware(concurrencyLimit int) *Middleware {
+	return &Middleware{
+		concurrencyLimit: int64(concurrencyLimit),
+	}
+}
+
+func (m *Middleware) RateLimitHandler(handler network.StreamHandler) network.StreamHandler {
 	return func(stream network.Stream) {
-		current := atomic.AddInt64(&parallelRequests, 1)
-		defer atomic.AddInt64(&parallelRequests, -1)
+		current := atomic.AddInt64(&m.parallelRequests, 1)
```

Comment:
Nit: iirc there's integer type for atomics now

Reply:
thanks, please check again